### PR TITLE
PR: Add the option to sync file order between the Outline Explorer and the current EditorStack

### LINKED ISSUE #7993: Order of root file items in the Outline Explorer should be synced with that of the current EditorStack
## Problem Description

As mentioned by @CAM-Gerlach in https://github.com/spyder-ide/spyder/pull/7968#discussion_r221500554, the outline explorer is still ordered alphabetically, rather than the order of the tabs in the Editor. It would be nice to have the option to either alphabetize it or make it follow the order of the editor, since both could potentially be useful depending on one's workflow.

This would be another step forward into making the Outline Explorer a viable proxy for a vertical tabbar for the Editor Stack.

@ccordoba12 @CAM-Gerlach

- Should we add an option in the context menu for this? Something like: `Show files sorted alphabetically` or something like that? Checking or unchecking this option would not impact the order of the files in the tabbar of the current EditorStack, only the order of the root file items in the Outline Explorer.

- Alternatively (or additionally), we could add an action in the context menu (with a corresponding button in the Outline Explorer toolbar) to actually sort the files alphabetically in the Outline Explorer and in the tabbar of the current EditorStack.

Also, do you think that one of the above suggestions could be a valid solution for Issue #5911?

### PR DESCRIPTION
**Pull request by jnsebgosselin**
### Issue(s) Resolved

Fixes #7993

## Description of Changes

- Added an option to toggle on/off sorting files alphabetically in the `Outline Explorer`
- When the above option is Unchecked, the files order are the same as in the tabbar of the current `EditorStack`.

sort_outline_explorer

### Pull Request Checklist

* [x] Added at least one unit test covering the changes, if at all possible
* [x] Described the changes and the motivation for them below
* [x] Included a screenshot, if this PR makes any visible changes to the UI

### Developer Certificate of Origin Affirmation

By submitting this Pull Request or typing my name below, I affirm the
[Developer Certificate of Origin](https://developercertificate.org/)
with respect to both the content of the contribution itself and this post,
and understand I am releasing it under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: Jean-Sébastien Gosselin

### CODE REVIEW

**Comment by jnsebgosselin** [comment]
@ccordoba12 Some tests are failing because of this. Do you think this is related to my work here:

`pip._vendor.pkg_resources.RequirementParseError: Invalid requirement, parse error at "'; extra '"`

**Comment by ccordoba12** [comment]
@jnsebgosselin, please merge with master to fix our tests.

**Comment by jnsebgosselin** [comment]
@ccordoba12 the test:
`spyder/utils/introspection/tests/test_jedi_plugin.py::test_get_info_from_method`
Do you think this is related to this work?

**Comment by ccordoba12** [comment]
That error should be fixed by PR #8004.

File: spyder/config/main.py [hunk 1]
```diff
@@ -269,6 +269,7 @@
               'show_fullpath': False,
               'show_all_files': False,
               'group_cells': True,
+              'show_files_sorted': False,
```
Comment (ccordoba12) [inline spyder/config/main.py]: What does this `sorted` mean? Alphabetically or by the Editor's order?

Comment (ccordoba12) [reply spyder/config/main.py]: I think this is a not so good choice of words that is applied to all your work in this PR and (IMHO) it should be changed.

Comment (jnsebgosselin) [reply spyder/config/main.py]: Alphabetically. I agree it is confusing. Do you have something in mind?

Comment (jnsebgosselin) [reply spyder/config/main.py]: I can rename the option to `show_files_in_alphabetical_order` or `show_files_sorted_alphabetically`, but it is a little bit long.

Comment (ccordoba12) [reply spyder/config/main.py]: What about `sort_files_alphabetically`?

Comment (CAM-Gerlach) [reply spyder/config/main.py]: Agreed as well. How about ``show_files_alphabetically``, ``show_files_alphabetized`` or even ``alphabetize_files``?

Comment (CAM-Gerlach) [reply spyder/config/main.py]: >What about ``sort_files_alphabetically``?

I like that best.

Comment (jnsebgosselin) [reply spyder/config/main.py]: I think I like `show_files_alphabetically` better than `sort_files_alphabetically`.

Maybe I am wrong about that, but the former add a nuance that this is not a permanent sorting we are doing and that this operation is only limited to what is shown in the Outline Explorer.

Comment (jnsebgosselin) [reply spyder/config/main.py]: Ok, majority wins, I'll use `sort_files_alphabetically` then.

Comment (CAM-Gerlach) [reply spyder/config/main.py]: Yeah, good point. Its very subtle, but if you think it could be confusing then yeah best be safe.

Comment (jnsebgosselin) [reply spyder/config/main.py]: I'll change the label of the action to `Sort files alphabetically` as well.

Comment (CAM-Gerlach) [reply spyder/config/main.py]: ^^

Comment (ccordoba12) [reply spyder/config/main.py]: Thanks @jnsebgosselin!

File: spyder/plugins/outlineexplorer/widgets.py [hunk 2]
```diff
@@ -620,10 +680,13 @@ def get_options(self):
                     show_all_files=self.treewidget.show_all_files,
                     group_cells=self.treewidget.group_cells,
                     show_comments=self.treewidget.show_comments,
+                    sort_files_alphabetically=\
+                        self.treewidget.sort_files_alphabetically,
                     expanded_state=self.treewidget.get_expanded_state(),
-                    scrollbar_position=self.treewidget.get_scrollbar_position(),
+                    scrollbar_position=\
```
Comment (CAM-Gerlach) [inline spyder/plugins/outlineexplorer/widgets.py]: PEP 8 and I believe our modern conventions these days strongly recommends never using ``\`` and always using ``()`` instead, and that's what's generally used in recently edited files.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 3]
```diff
@@ -620,10 +680,13 @@ def get_options(self):
                     show_all_files=self.treewidget.show_all_files,
                     group_cells=self.treewidget.group_cells,
                     show_comments=self.treewidget.show_comments,
+                    sort_files_alphabetically=\
```
Comment (CAM-Gerlach) [inline spyder/plugins/outlineexplorer/widgets.py]: See PEP8/our conventions

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 4]
```diff
@@ -35,16 +35,16 @@ def test_files(tmpdir_factory):
...
-        f.write("# -*- coding: utf-8 -*-\n"
-                "# ---- a comment\n")
-
-    filename3 = osp.join(tmpdir.strpath, 'text1.txt')
-    with open(filename3, 'w') as f:
         f.write("This is a simple text file for\n"
                 "testing the Outline Explorer.\n")
 
+    filename3 = osp.join(tmpdir.strpath, 'foo2.py')
+    with open(filename3, 'w') as f:
```
Comment (CAM-Gerlach) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: You should add ``encoding="utf-8" for consistency across platforms and general standard conformance (since on Windows this will default to the locale encoding, which could theoretically be anything and may or may not read the file correctly or completly), especially since you define the file encoding as UTF-8 below.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 5]
```diff
@@ -35,16 +35,16 @@ def test_files(tmpdir_factory):
                 "def foo:\n"
                 "    print(Hello World!)\n")
 
-    filename2 = osp.join(tmpdir.strpath, 'foo2.py')
+    filename2 = osp.join(tmpdir.strpath, 'text1.txt')
     with open(filename2, 'w') as f:
```
Comment (CAM-Gerlach) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: Add ``encoding="utf-8"`` here as discussed below.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 6]
```diff
@@ -138,7 +138,80 @@ def test_close_a_file(editorstack, outlineexplorer, test_files):
...
+
+    This feature was introduced in PR #8015
+    """
+    editorstack = editorstack(test_files)
+    treewidget = outlineexplorer.treewidget
+    results = [item.text(0) for item in treewidget.get_visible_items()]
+    assert results == ['foo1.py', 'text1.txt', 'foo2.py']
+
+    # Set the option to sort files alphabetically to True and assert the
+    # results.
```
Comment (CAM-Gerlach) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: "Assert the results"?

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 7]
```diff
@@ -138,7 +138,80 @@ def test_close_a_file(editorstack, outlineexplorer, test_files):
...
+    tabs are moved in the EditorStack.
+
+    This feature was introduced in PR #8015
+    """
+    editorstack = editorstack(test_files)
+    treewidget = outlineexplorer.treewidget
+    results = [item.text(0) for item in treewidget.get_visible_items()]
+    assert results == ['foo1.py', 'text1.txt', 'foo2.py']
+
+    # Invert tab 1 with tab 2 and assert the results.
```
Comment (CAM-Gerlach) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: "Assert the results"?

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 8]
```diff
@@ -138,7 +138,80 @@ def test_close_a_file(editorstack, outlineexplorer, test_files):
...
+    when the current Editor in the Editorstack changes.
+
+    Test that toggling back the option to show all files, after the
+    order of the files in the Editorstack was changed, show all the root
+    file items in the correct order.
+    """
+    editorstack = editorstack(test_files)
+    treewidget = outlineexplorer.treewidget
+
+    # Untoggle show all files option and assert the results.
```
Comment (CAM-Gerlach) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: Assert the results

File: spyder/plugins/outlineexplorer/widgets.py [hunk 9]
```diff
@@ -310,11 +330,47 @@ def remove_editor(self, editor):
                 except RuntimeError:
                     # item has already been removed
                     pass
-        
+
+    def set_editor_ids_order(self, ordered_editor_ids):
+        """
+        Order the root files items in the Outline Explorer following the
```
Comment (CAM-Gerlach) [inline spyder/plugins/outlineexplorer/widgets.py]: If we're not going to follow PEP 257, we shouldn't advertise that we do in our CONTRIBUTING guide.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 10]
```diff
@@ -310,11 +330,47 @@ def remove_editor(self, editor):
...
+            self.ordered_editor_ids = ordered_editor_ids
+            if self.show_all_files and self.sort_files_alphabetically is False:
+                self.__sort_toplevel_items()
+
     def __sort_toplevel_items(self):
-        sort_func = lambda item: osp.basename(item.path.lower())
-        self.sort_top_level_items(key=sort_func)
-            
+        """
+        Sort the root file items in alphabetical order if
```
Comment (CAM-Gerlach) [inline spyder/plugins/outlineexplorer/widgets.py]: See PEP 257 comment

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py (hunk 7 above)
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: Are you suggesting that I remove "Assert the results" from the comments? I do not understand what you are suggesting.

File: spyder/plugins/outlineexplorer/widgets.py (hunk 9 above)
Comment (ccordoba12) [reply spyder/plugins/outlineexplorer/widgets.py]: We agreed not to enforce it. So it's up to the contributor to follow it or not.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 11]
```diff
@@ -611,19 +671,23 @@ def set_current_editor(self, editor, update, clear):
...
-                    expanded_state=self.treewidget.get_expanded_state(),
-                    scrollbar_position=self.treewidget.get_scrollbar_position(),
-                    visibility=self.isVisible())
-    
+        return dict(
+            show_fullpath=self.treewidget.show_fullpath,
+            show_all_files=self.treewidget.show_all_files,
+            group_cells=self.treewidget.group_cells,
+            show_comments=self.treewidget.show_comments,
+            sort_files_alphabetically=
```
Comment (jnsebgosselin) [inline spyder/plugins/outlineexplorer/widgets.py]: I do not know how to fix the two pep8 issues related to these two lines elegantly. Please help.

**Comment by CAM-Gerlach** [comment]
@jnsebgosselin Also, not sure if its related, but just now I noticed when hitting Shift-Tab to try to help with your PEP 8 issue, the tab is removed but then the cursor switches to the other Editor pane, at least after triggering the error above with inconsistent files open on each one, even after making the files consistent between the two. It doesn't happen in a new pane I open, but it does in the existing pane (the one that I didn't do the "close all" command, that had some inconsistant files left open, and I can't close it this time since it happens to be the "first" pane.

**Comment by CAM-Gerlach** [comment]
@jnsebgosselin Actually, this seems to be unrelated as it reoccured when I closed and re-opened Spyder; it seems to happen at random in opened panes where Shift-Tab is not "trapped" by the Editor window, and doing Shift-Tab will delete the tabs but then move the tab focus to the Spyder UI. Maybe this has to do with your Shift keyboard shortcuts PR? Tab doesn't do it, only Shift-Tab.

File: spyder/plugins/outlineexplorer/widgets.py (hunk 11 above)
Comment (CAM-Gerlach) [reply spyder/plugins/outlineexplorer/widgets.py]: Its just

```python
sort_files_alphabetically=(
    self.treewidget.sort_files_alphabetically)
```

**Comment by jnsebgosselin** [comment]
> @jnsebgosselin Actually, this seems to be unrelated as it reoccured when I closed and re-opened Spyder; it seems to happen at random in opened panes where Shift-Tab is not "trapped" by the Editor window, and doing Shift-Tab will delete the tabs but then move the tab focus to the Spyder UI. Maybe this has to do with your Shift keyboard shortcuts PR? Tab doesn't do it, only Shift-Tab.

You are right. This was not introduced by a change I made since this behavior is also present in the Beta1 and I started to work on shortcuts after this was release.

`Shift-Tab` action is currently hard-coded in the Editor. I will take a look and see if I can do something about this.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py (hunk 4 above)
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: This is not Python 2.7 friendly it seems...

Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: ... so reverting since everything is ASCII in those test files anyway.

File: spyder/plugins/outlineexplorer/widgets.py (hunk 11 above)
Comment (jnsebgosselin) [reply spyder/plugins/outlineexplorer/widgets.py]: Cool, thanks

**Comment by jnsebgosselin** [comment]
> If I toggle or untoggle "Sort files alphabetically", a list of all open files will appear regardless of whether "Show all files" is checked. It will go back to showing only the current file under the same conditions as the above bug.

Fixed by https://github.com/spyder-ide/spyder/pull/8015/commits/58de2d8ba969c67cfec88c119ade81e606e11181

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py (hunk 4 above)
Comment (CAM-Gerlach) [reply spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: Ah. Looking it up, apparently the best practice is to just always use ``io.open()`` instead of open() if needed Python 2 compatible code, as it includes an encoding argument and works in Python 2 and Python 3 identically; in fact, in Python 3, open() is just an alias for io.open(). This ensures compatibility between the two versions and ensures files will be decoded and encoded consistently and with full Unicode compatibility.

**Comment by jnsebgosselin** [comment]
>When I open files with "Show all files" turned off, the last file loaded is shown in the outline explorer, even if it isn't the one opened in either pane. However, it quickly updates itself to the current file once I change files or click in the other pane, so its just a minor transient cosmetic bug.

This is unrelated to this PR, but a bug nevertheless. This is in fact related to #7338. Dood practice tells me I should fix this in another PR, but I'm tempted to be practical and ninja fix this here lol.

**Comment by jnsebgosselin** [comment]
> When I click "close all to right" or "close all but this" on a file in the one split pane when the file order in the EditorStack is not alphabetical (regardless of whether it is identical between panes, which pane (or both) the file was moved on, the outline explorer is alphabetical mode, show all files is on or not, and which pane "close all" is selected on, all of which I tested), all the files except for the first close on the pane I clicked "close tall to right" on (as they should) while any files that I've moved, as well as sometimes the files higher in the alphabet than the moved file, don't close on the other pane, and (if alphabetical mode is initially checked) I see duplicate copies of each file (which is resolved when I toggle the alphabetical option at least once, but not when switching files). Furthermore, I get the following error:

Good catch. I think https://github.com/spyder-ide/spyder/pull/8015/commits/ea53bd392e209d7d78216ad7a53254a40dc8a318 should have taken care of this error.

**Comment by jnsebgosselin** [comment]
> 
> 
> > @jnsebgosselin Actually, this seems to be unrelated as it reoccured when I closed and re-opened Spyder; it seems to happen at random in opened panes where Shift-Tab is not "trapped" by the Editor window, and doing Shift-Tab will delete the tabs but then move the tab focus to the Spyder UI. Maybe this has to do with your Shift keyboard shortcuts PR? Tab doesn't do it, only Shift-Tab.
> 
> You are right. This was not introduced by a change I made since this behavior is also present in the Beta1 and I started to work on shortcuts after this was released.
> 
> `Shift-Tab` action is currently hard-coded in the Editor. I will take a look and see if I can do something about this.

@CAM-Gerlach this bug needs its own issue. Do you mind creating one please?

**Comment by jnsebgosselin** [comment]
> 
> 
> > > When I open files with "Show all files" turned off, the last file loaded is shown in the outline explorer, even if it isn't the one opened in either pane. However, it quickly updates itself to the current file once I change files or click in the other pane, so its just a minor transient cosmetic bug.
> > 
> > 
> > This is unrelated to this PR, but a bug nevertheless. This is in fact related to #7338. Good practice tells me I should fix this in another PR, but I'm tempted to be practical and ninja fix this here lol.
> 
> Finally, I decided to fix this in commit [10d0bde](https://github.com/spyder-ide/spyder/commit/10d0bde9028a9c280c79ea669df381645be5b441). Also, this solve the flickering issue mentioned in [#7968 (comment)](https://github.com/spyder-ide/spyder/pull/7968#discussion_r221490842)

This doesn't work as expected... working to fix this and I will expand the tests to cover these corner cases.

**Comment by jnsebgosselin** [comment]
> 
> 
> > @CAM-Gerlach this bug needs its own issue. Do you mind creating one please?
> 
> Sure thing; done. Sorry for the delay; I had some medical issues with my cats yesterday I had to take care of.

Thank you very much. I do not consider this a delay hehe. I have 3 cats also, so I know what you are talking about.

@ccordoba12 this is **not** ready to merge. Considering there is a lot of corner cases to consider, I want to test it a little bit further. You are welcome to review however if you want and have the time.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 12]
```diff
@@ -137,9 +138,100 @@ def test_close_a_file(editorstack, outlineexplorer, test_files):
...
+
+    # Switch tab 1 with tab 2.
+    editorstack.tabs.tabBar().moveTab(0, 1)
+    results = [item.text(0) for item in treewidget.get_visible_items()]
+    assert results == ['text1.txt', 'foo1.py', 'foo2.py']
+
+
+# ---- Test single file mode
+def test_toggle_off_show_all_files(editorstack, outlineexplorer,
+                                   test_files, qtbot):
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: `qtbot` is not needed in this test.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 13]
```diff
@@ -137,9 +138,100 @@ def test_close_a_file(editorstack, outlineexplorer, test_files):
...
+
+    # Select the last file in the Editorstack.
+    with qtbot.waitSignal(editorstack.editor_focus_changed):
+        editorstack.tabs.setCurrentIndex(2)
+    results = [item.text(0) for item in treewidget.get_visible_items()]
+    assert results == ['foo2.py', '---- a comment']
+
+
+def test_toggle_on_show_all_files(editorstack, outlineexplorer,
+                                  test_files, qtbot):
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py]: `qtbot` is also not needed here.

**Comment by jnsebgosselin** [comment]
Excellent thanks

**Comment by jnsebgosselin** [comment]
Thank you very much @CAM-Gerlach for your time reviewing this, this is much appreciated.

**Comment by jnsebgosselin** [comment]
I forgot to increase the conf version. I'll do this right now.

**Comment by jnsebgosselin** [comment]
> You don't need to touch this value if you're just adding a new option

Ah! I don't need to do anything. So all is good.
